### DISCUSSION TITLE
GCC compiler options difficulty

### DISCUSSION THREAD

**Discussion by dominik-mokris-mtu** [gismo/gismo#688] [Announcements]
**Update: a solution is [below](https://github.com/gismo/gismo/discussions/688#discussioncomment-9285955).**

Today, I would like to share with you my recent difficulties with gcc options, because it is not the first time I ran into them, cf. https://github.com/gismo/gismo/discussions/566#discussioncomment-4381047 .

When I run ``CMake`` for the first time, the ``gismo/cmake/OptimizeForArchitecture.cmake`` script attempts quite vigorously to find the compiler settings to squeeze the best possible performance out of the current CPU. On my computer, the result typically looks like this
```
> make VERBOSE=1
[...]
[  0%] Building CXX object src/CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o
cd /path/to/gismo/tdm/tst/src && /usr/bin/c++ -Dgismo_EXPORTS -I/path/to/gismo/tdm/gismo/src -I/path/to/gismo/tdm/gismo/external -I/path/to/gismo/tdm/tst -I/path/to/gismo/tdm/gismo/optional -Wall -Wno-long-long -Wunused-variable -Wno-unknown-pragmas -march=skylake-avx512 -mmmx -mavx -mavx2 -mfma -msse2 -msse3 -msse4.1 -msse4.2 -msse -mssse3 -mavx512bw -mavx512cd -mavx512dq -mavx512f -mavx512vl -maes -mbmi2 -mf16c -mfsgsbase -mpclmul -mpku -mpopcnt -mrdrnd -mabm -mbmi -mno-sse4a -mno-avx5124fmaps -mno-avx5124vnniw -mno-avx512er -mno-avx512ifma -mno-avx512pf -mno-avx512vbmi -mno-avx512vpopcntdq -mno-adx -mno-fxsr -mno-lzcnt -mno-movbe -mno-mpx -mno-prfchw -mno-prefetchwt1 -mno-rdpid -mno-rdseed -mno-rtm -mno-sha -mno-xsavec -mno-xsaveopt -mno-xsaves -mno-clflushopt -mno-clwb -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -std=c++14 -MD -MT src/CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o -MF CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o.d -o CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o -c /path/to/gismo/tdm/gismo/src/gsParallel/gsOpenMP.cpp
```

For reasons out of my direct control, the CPU information passed to the script is not 100% accurate in our environment and it can activate unsuitable flags. In practice this means that call to G+Smo that worked perfectly well two weeks ago can suddenly start crashing with ``signal 4 (Illegal instruction)`` in a seemingly random spot.

What's worse, the problem is difficult to debug: re-compiling in ``Debug`` mode and using GDB finds no problem at all; in fact, the program runs perfectly well, because the optimization script is active only in ``Release`` mode. This leaves you guessing in the dark.

A dear colleague of mine helped me find a way around this: start with a fresh build-folder and configure with
```
> cmake /path/to/gismo -DARCHITECTURE_CXX_FLAGS=BAKE_MORE_CAKE
```
Setting the ``CMake`` variable ``ARCHITECTURE_CXX_FLAGS`` to anything (it does not seem to matter what) deactivates the script (cf. ``gismo/cmake/gsConfig.cmake``), leading to slower but more reliable code. This can be again checked when asking for a verbose ``make``:
```
> make VERBOSE=1
[...]
[  0%] Building CXX object src/CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o
cd /path/to/gismo/tdm/tst/src && /usr/bin/c++ -Dgismo_EXPORTS -I/path/to/gismo/tdm/gismo/src -I/path/to/gismo/tdm/gismo/external -I/path/to/gismo/tdm/tst -I/path/to/gismo/tdm/gismo/optional -Wall -Wno-long-long -Wunused-variable -Wno-unknown-pragmas -Wall -Wno-long-long -Wunused-variable -Wno-unknown-pragmas -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -std=c++14 -MD -MT src/CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o -MF CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o.d -o CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o -c /path/to/gismo/tdm/gismo/src/gsParallel/gsOpenMP.cpp
```

Any thoughts? I guess that @mmoelle1 has quite some experience in the area.

**Comment by mmoelle1** [comment]
Let’s discuss this in a zoom call next week?

Am 15.02.2024 um 18:08 schrieb Dominik Mokriš - work account ***@***.***>:

﻿

Today, I would like to share with you my recent difficulties with gcc options, because it is not the first time I ran into them.

When I run CMake for the first time, the gismo/cmake/OptimizeForArchitecture.cmake script attempts quite vigorously to find the compiler settings to squeeze the best possible performance out of the current CPU. On my computer, the result typically looks like this

> make VERBOSE=1
[...]
[  0%] Building CXX object src/CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o
cd /path/to/gismo/tdm/tst/src && /usr/bin/c++ -Dgismo_EXPORTS -I/path/to/gismo/tdm/gismo/src -I/path/to/gismo/tdm/gismo/external -I/path/to/gismo/tdm/tst -I/path/to/gismo/tdm/gismo/optional -Wall -Wno-long-long -Wunused-variable -Wno-unknown-pragmas -march=skylake-avx512 -mmmx -mavx -mavx2 -mfma -msse2 -msse3 -msse4.1 -msse4.2 -msse -mssse3 -mavx512bw -mavx512cd -mavx512dq -mavx512f -mavx512vl -maes -mbmi2 -mf16c -mfsgsbase -mpclmul -mpku -mpopcnt -mrdrnd -mabm -mbmi -mno-sse4a -mno-avx5124fmaps -mno-avx5124vnniw -mno-avx512er -mno-avx512ifma -mno-avx512pf -mno-avx512vbmi -mno-avx512vpopcntdq -mno-adx -mno-fxsr -mno-lzcnt -mno-movbe -mno-mpx -mno-prfchw -mno-prefetchwt1 -mno-rdpid -mno-rdseed -mno-rtm -mno-sha -mno-xsavec -mno-xsaveopt -mno-xsaves -mno-clflushopt -mno-clwb -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -std=c++14 -MD -MT src/CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o -MF CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o.d -o CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o -c /path/to/gismo/tdm/gismo/src/gsParallel/gsOpenMP.cpp


For reasons out of my direct control, the CPU information passed to the script is not 100% accurate in our environment and it can activate unsuitable flags. In practice this means that call to G+Smo that worked perfectly well two weeks ago can suddenly start crashing with signal 4 (Illegal instruction) in a seemingly random spot.

What's worse, the problem is difficult to debug: re-compiling in Debug mode and using GDB finds no problem at all; in fact, the program runs perfectly well, because the optimization script is active only in Release mode. This leaves you guessing in the dark.

A dear colleague of mine helped me find a way around this: start with a fresh build-folder and configure with

> cmake /path/to/gismo -DARCHITECTURE_CXX_FLAGS=BAKE_MORE_CAKE


Setting the CMake variable ARCHITECTURE_CXX_FLAGS to anything (it does not seem to matter what) deactivates the script (cf. gismo/cmake/gsConfig.cmake), leading to slower but more reliable code. This can be again checked when asking for a verbose make:

> make VERBOSE=1
[...]
[  0%] Building CXX object src/CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o
cd /path/to/gismo/tdm/tst/src && /usr/bin/c++ -Dgismo_EXPORTS -I/path/to/gismo/tdm/gismo/src -I/path/to/gismo/tdm/gismo/external -I/path/to/gismo/tdm/tst -I/path/to/gismo/tdm/gismo/optional -Wall -Wno-long-long -Wunused-variable -Wno-unknown-pragmas -Wall -Wno-long-long -Wunused-variable -Wno-unknown-pragmas -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -std=c++14 -MD -MT src/CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o -MF CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o.d -o CMakeFiles/gsParallel.dir/gsParallel/gsOpenMP.cpp.o -c /path/to/gismo/tdm/gismo/src/gsParallel/gsOpenMP.cpp


Any thoughts? I guess that @mmoelle1<https://urldefense.com/v3/__https://github.com/mmoelle1__;!!PAKc-5URQlI!7lgUiLoAcK1vTFwd-2uEVhs-mvvkjSK3Cb0xUmlSdNU7fh3fXLG5Qj0M0deB--EHtVahW9v3bx1t9dwP4YVWpIvqWag$> has quite some experience in the area.

**Reply by dominik-mokris-mtu** [reply]
Hi Matthias,

I am sorry for not responding earlier: there were more pressing topics.

In the meantime I have done a few experiments and collected additional relevant information. If you still want to have a look at the problem together, we can try to set a meeting in the upcoming weeks or wait until the next developer call.

**Comment by mmoelle1** [comment]
Hi Dominik,

yes, I’m still interested to fix this bug. We could meet 23/04 at 17:00.



Am 12.04.2024 um 12:08 schrieb Dominik Mokriš - work account ***@***.***>:

﻿

Hi Matthias,

I am sorry for not responding earlier: there were more pressing topics.

In the meantime I have done a few experiments and collected additional relevant information. If you still want to have a look at the problem together, we can try to set a meeting in the upcoming weeks or wait until the next developer call.

**Comment by dominik-mokris-mtu** [comment]
After quite some time spent learning, discussing and testing I have found where I had made the mistake. I share it here in the hope of helping the others to avoid the same pitfall.

The problem, which was manifesting itself with the `segmentation fault`s, indeed had to do with the optimization settings of gcc. The `OptimizeForArchitecture.cmake` correctly identified my processor and tried to find the best settings for it; however, in my case (big multi-user Linux environment) it sometimes had to run on _different_ processors as well. These might have slightly different instruction sets, leading to the crash.

Very similar problem would happen even when simply setting `-O3` in gcc (as is the case in the `Release` mode of G+Smo): newer gcc versions then use vectorization and the compiled object is not necessarily transferable to other processors.

A solution is described in `OptimizeForArchitecture.cmake` as well: set `-DTARGET_ARCHITECTURE=generic` in CMake. This makes sure that CPU-specific optimizations are de-activated. Personally, I would prefer to explicitly set the architecture to `x86-64` (I would choose this when setting directly the `-march` [gcc option](https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html) now that I know what processor class I need to support) but as it does not seem to be available, I am perfectly fine with `generic`.

Out of curiosity, I did some benchmarking, measuring the runtimes of `fitting_example.cpp` with a bigger point cloud three times:
```
|----------------+---------------------+-------------+-------------+-------------|
| build type     | TARGET_ARCHITECTURE | time (real) | time (real) | time (real) |
|----------------+---------------------+-------------+-------------+-------------|
| RelWithDebinfo | auto                | 0m20.935s   | 0m20.762s   | 0m20.887s   |
| Release        | auto                | 0m13.969s   | 0m14.528s   | 0m14.065s   |
| Release        | generic             | 0m13.930s   | 0m13.872s   | 0m13.894s   |
| Release        | none                | 0m12.558s   | 0m12.530s   | 0m12.585s   |
|----------------+---------------------+-------------+-------------+-------------|
```
This was surprising for me: I would have expected the `Release` runtimes to be `auto` < `generic` < `none` but that is not the case, at least in this example.

HTH!

**Reply by dominik-mokris-mtu** [reply]
Related: #374 , #450.

**Reply by mmoelle1** [reply]
We can change `generic` to `x86_64` but then we would need a different "generic" for different platforms, e.g., `ppc64` and `aarch64`. I think that a platform-independent `generic` is most generic :-)

Concerning the performance results. The influence of the optimization depends strongly on the type of operations your application uses most. I will do the same benchmark with another application and report them here. BTW, at the moment `RelWithDebInfo` does not use any optimization, so it should be the same as `none`. Are you sure that you removed the build folder before rebuilding the benchmark application?

Results of the `poisson2_example` benchmark 

**auto**: CMake configuration: `CC=gcc CXX=g++ cmake .. -DTARGET_ARCHITECTURE=auto -DCMAKE_BUILD_TYPE=Release`

The identified architecture is `kabylake` with the following optimization flags:
```
-march=skylake;-mmmx;-mavx;-mavx2;-mfma;-msse2;-msse3;-msse4.1;-msse4.2;-msse;-mssse3;-madx;-maes;-mbmi2;-mf16c;-mfsgsbase;-mpclmul;-mpopcnt;-mprfchw;-mrdrnd;-mrdseed;-mxsavec;-mxsaves;-mabm;-mbmi;-mclflushopt;-mno-sse4a;-mno-avxvnni;-mno-avx5124fmaps;-mno-avx5124vnniw;-mno-avx512bf16;-mno-avx512bitalg;-mno-avx512bw;-mno-avx512cd;-mno-avx512dq;-mno-avx512er;-mno-avx512f;-mno-avx512fp16;-mno-avx512ifma;-mno-avx512pf;-mno-avx512vbmi2;-mno-avx512vbmi;-mno-avx512vl;-mno-avx512vnni;-mno-avx512vp2intersect;-mno-avx512vpopcntdq;-mno-amx-bf16;-mno-amx-int8;-mno-amx-tile;-mno-enqcmd;-mno-fxsr;-mno-gfni;-mno-hreset;-mno-lzcnt;-mno-movbe;-mno-movdir64b;-mno-movdiri;-mno-mpx;-mno-pconfig;-mno-pku;-mno-prefetchwt1;-mno-ptwrite;-mno-rdpid;-mno-rtm;-mno-serialize;-mno-sha;-mno-tsxldtrk;-mno-uintr;-mno-vaes;-mno-vpclmulqdq;-mno-waitpkg;-mno-wbnoinvd;-mno-xsaveopt;-mno-cldemote;-mno-clwb
```

**generic**: CMake configuration: `CC=gcc CXX=g++ cmake .. -DTARGET_ARCHITECTURE=generic -DCMAKE_BUILD_TYPE=Release`

The identified architecture is `generic` with the following optimization flags:
```
-msse2;-msse;-mno-mmx;-mno-avx;-mno-avx2;-mno-fma;-mno-sse3;-mno-sse4.1;-mno-sse4.2;-mno-sse4a;-mno-ssse3;-mno-avxvnni;-mno-avx5124fmaps;-mno-avx5124vnniw;-mno-avx512bf16;-mno-avx512bitalg;-mno-avx512bw;-mno-avx512cd;-mno-avx512dq;-mno-avx512er;-mno-avx512f;-mno-avx512fp16;-mno-avx512ifma;-mno-avx512pf;-mno-avx512vbmi2;-mno-avx512vbmi;-mno-avx512vl;-mno-avx512vnni;-mno-avx512vp2intersect;-mno-avx512vpopcntdq;-mno-amx-bf16;-mno-amx-int8;-mno-amx-tile;-mno-adx;-mno-aes;-mno-bmi2;-mno-enqcmd;-mno-f16c;-mno-fsgsbase;-mno-fxsr;-mno-gfni;-mno-hreset;-mno-lzcnt;-mno-movbe;-mno-movdir64b;-mno-movdiri;-mno-mpx;-mno-pclmul;-mno-pconfig;-mno-pku;-mno-popcnt;-mno-prfchw;-mno-prefetchwt1;-mno-ptwrite;-mno-rdpid;-mno-rdrnd;-mno-rdseed;-mno-rtm;-mno-serialize;-mno-sha;-mno-tsxldtrk;-mno-uintr;-mno-vaes;-mno-vpclmulqdq;-mno-waitpkg;-mno-wbnoinvd;-mno-xsavec;-mno-xsaveopt;-mno-xsaves;-mno-abm;-mno-bmi;-mno-cldemote;-mno-clflushopt;-mno-clwb
```

**none**: CMake configuration: `CC=gcc CXX=g++ cmake .. -DTARGET_ARCHITECTURE=none -DCMAKE_BUILD_TYPE=Release`

The identified architecture is `none` with the following optimization flags: N/A

**Running**: `time OMP_NUM_THREADS=1 ./bin/poisson2_example -r 6 -e 3`

I have set OpenMP to 1 thread since using the "wrong" number of OpenMP threads can also lead to performance decrease.

|----------------+---------------------+-------------+-------------+-------------|
| build type     | TARGET_ARCHITECTURE | time (real) | time (real) | time (real) |
|----------------+---------------------+-------------+-------------+-------------|
| Release        | auto                | 0m14.383s | 0m14.467s | 0m14.441s |
| Release        | generic             | 0m15.682s | 0m15.458s | 0m15.244s |
| Release        | none                | 0m15.574s | 0m15.537s | 0m15.170s |
|----------------+---------------------+-------------+-------------+-------------|

**Reply by dominik-mokris-mtu** [reply]
> We can change generic to x86_64 but then we would need a different "generic" for different platforms, e.g., ppc64 and aarch64. I think that a platform-independent generic is most generic :-)

Exactly. What I had in mind would be an `x86_64` option somewhere between `generic` and `auto`, analogously to how the `-mtune` [gcc option](https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html) is structured. But I definitely agree that it does not seem to be worth the effort now.

> at the moment RelWithDebInfo does not use any optimization, so it should be the same as none.

I don't think so. `RelWithDebInfo` translates to `-O2 -g`, which is something else than `-O3` with `-DTARGET_ARCHITECTURE=none`, isn't it?

> Are you sure that you removed the build folder before rebuilding the benchmark application?

I have a separate option for each of the configurations but I will check once again to be sure. And I will report later what the _poisson2_example_ runtimes are.

**Reply by mmoelle1** [reply]
Concerning the `x86_64` option. We could define `avx`, `avx2` and `avx512` as new architectures and map them to the first processor that supports these extensions. With `avx512` that would be tricky because there is not "the one and only" `avx512` extension. What I typically do when I want to generate an optimised "universal" build, e.g., in docker is that I use `nehalem` (pre `avx` but `ss(s)e`), `sandybridge` (= `avx`), `haskell` (= `avx2`), `skylake_avx512` (= first `avx512` extensions).

**Reply by dominik-mokris-mtu** [reply]
> Results of the poisson2_example benchmark

For the sake of completeness, I have done similar benchmarking.

## `TARGET_ARCHITECTURE` in CMake
I used the same CMake configuration as Matthias, except specifically taking gcc 11.2. Below are excerpts from CMake outputs for various `TARGET_ARCHITECTURE` values. Each is done in a separate folder in order to build from scratch.

### auto
```
-- Target architecture changed from "" to "auto"
-- Detected Host CPU: zen2
-- Checking Host CPU features. This can take some time ...
-- (!) x86_64 architecture optimization flags applied: -march=znver2;-mmmx;-mavx;-mavx2;-mfma;-msse2;-msse3;-msse4.1;-msse4.2;-msse4a;-msse;-mssse3;-maes;-mbmi2;-mf16c;-mfsgsbase;-mpclmul;-mpopcnt;-mrdpid;-mrdseed;-msha;-mwbnoinvd;-mxsavec;-mxsaves;-mabm;-mbmi;-mclflushopt;-mclwb;-mno-avxvnni;-mno-avx5124fmaps;-mno-avx5124vnniw;-mno-avx512bf16;-mno-avx512bitalg;-mno-avx512bw;-mno-avx512cd;-mno-avx512dq;-mno-avx512er;-mno-avx512f;-mno-avx512ifma;-mno-avx512pf;-mno-avx512vbmi2;-mno-avx512vbmi;-mno-avx512vl;-mno-avx512vnni;-mno-avx512vp2intersect;-mno-avx512vpopcntdq;-mno-amx-bf16;-mno-amx-int8;-mno-amx-tile;-mno-adx;-mno-enqcmd;-mno-fxsr;-mno-gfni;-mno-hreset;-mno-lzcnt;-mno-movbe;-mno-movdir64b;-mno-movdiri;-mno-mpx;-mno-pconfig;-mno-pku;-mno-prfchw;-mno-prefetchwt1;-mno-ptwrite;-mno-rdrnd;-mno-rtm;-mno-serialize;-mno-tsxldtrk;-mno-uintr;-mno-vaes;-mno-vpclmulqdq;-mno-waitpkg;-mno-xsaveopt;-mno-cldemote.
```

### generic
```
-- Target architecture changed from "" to "generic"
-- Checking Host CPU features. This can take some time ...
-- (!) x86_64 architecture optimization flags applied: -msse;-mno-mmx;-mno-avx;-mno-avx2;-mno-fma;-mno-sse2;-mno-sse3;-mno-sse4.1;-mno-sse4.2;-mno-sse4a;-mno-ssse3;-mno-avxvnni;-mno-avx5124fmaps;-mno-avx5124vnniw;-mno-avx512bf16;-mno-avx512bitalg;-mno-avx512bw;-mno-avx512cd;-mno-avx512dq;-mno-avx512er;-mno-avx512f;-mno-avx512ifma;-mno-avx512pf;-mno-avx512vbmi2;-mno-avx512vbmi;-mno-avx512vl;-mno-avx512vnni;-mno-avx512vp2intersect;-mno-avx512vpopcntdq;-mno-amx-bf16;-mno-amx-int8;-mno-amx-tile;-mno-adx;-mno-aes;-mno-bmi2;-mno-enqcmd;-mno-f16c;-mno-fsgsbase;-mno-fxsr;-mno-gfni;-mno-hreset;-mno-lzcnt;-mno-movbe;-mno-movdir64b;-mno-movdiri;-mno-mpx;-mno-pclmul;-mno-pconfig;-mno-pku;-mno-popcnt;-mno-prfchw;-mno-prefetchwt1;-mno-ptwrite;-mno-rdpid;-mno-rdrnd;-mno-rdseed;-mno-rtm;-mno-serialize;-mno-sha;-mno-tsxldtrk;-mno-uintr;-mno-vaes;-mno-vpclmulqdq;-mno-waitpkg;-mno-wbnoinvd;-mno-xsavec;-mno-xsaveopt;-mno-xsaves;-mno-abm;-mno-bmi;-mno-cldemote;-mno-clflushopt;-mno-clwb.
```

### none
```
-- Target architecture changed from "" to "none"
-- Checking Host CPU features. This can take some time ...
```

## Runtimes
### fitting_example
This is my previous benchmark (18351 points) with `OMP_NUM_THREADS` as suggested.
**call:** `time OMP_NUM_THREADS=1 ./bin/fitting_example -c 0 -i 6 -x 3 -y 3 -s 1e-8 -p 0.95 -q 2 -r 2 -e 1e-5 -d data.xml`
```
|----------------+---------------------+-------------+-------------+-------------|
| build type     | TARGET_ARCHITECTURE | time (real) | time (real) | time (real) |
|----------------+---------------------+-------------+-------------+-------------|
| RelWithDebInfo |                     | 0m21.123s   | 0m21.050s   | 0m21.285s   |
| Release        | auto                | 0m12.824s   | 0m12.841s   | 0m12.800s   |
| Release        | generic             | 0m13.950s   | 0m13.946s   | 0m13.915s   |
| Release        | none                | 0m12.718s   | 0m12.713s   | 0m12.647s   |
|----------------+---------------------+-------------+-------------+-------------|
```
The runtimes are similar to the previous results except accelerating `auto`, in line with Matthias' expectation.

### poisson2_example
**call:** `time OMP_NUM_THREADS=1 ./bin/poisson2_example -r 6 -e 3`
```
|----------------+---------------------+-------------+-------------+-------------|
| build type     | TARGET_ARCHITECTURE | time (real) | time (real) | time (real) |
|----------------+---------------------+-------------+-------------+-------------|
| RelWithDebInfo |                     | 0m27.724s   | 0m27.797s   | 0m27.834s   |
| Release        | auto                | 0m16.720s   | 0m16.938s   | 0m16.755s   |
| Release        | generic             | 0m28.584s   | 0m28.544s   | 0m28.822s   |
| Release        | none                | 0m18.754s   | 0m18.783s   | 0m18.639s   |
|----------------+---------------------+-------------+-------------+-------------|
```
I have also tried the same without specifying `OMP_NUM_THREADS` but there was no significant change.

## Conclusions
With `OMP_NUM_THREADS=1` the runtimes are `auto` < `none` < `generic`. This sounds reasonable: if `generic` explicitly turns off vectorisation, then it should be slower than `none`, which can still vectorise due to `-O3`. I suppose that the speed-up of `auto` would be more pronounced on bigger examples.

From my viewpoint we don't need to change anything in code. My take-home lesson is to use `generic` if I need the shared object to be portable.

**Comment by dominik-mokris-mtu** [comment]
> Concerning the `x86_64` option. We could define `avx`, `avx2` and `avx512` as new architectures and map them to the first processor that supports these extensions. With `avx512` that would be tricky because there is not "the one and only" `avx512` extension. What I typically do when I want to generate an optimised "universal" build, e.g., in docker is that I use `nehalem` (pre `avx` but `ss(s)e`), `sandybridge` (= `avx`), `haskell` (= `avx2`), `skylake_avx512` (= first `avx512` extensions).

Personally, I think this is not worth the effort. The `generic` option seems safe and fast enough in my situation. Thanks for the ideas nonetheless! (-:
